core/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java

```diff
@@ -75,30 +82,89 @@ private IntegrationStep step(double timeStep, double position, double speed) {
 
         double tractionForce = 0;
         double brakingForce = 0;
-        double maxTractionForce = rollingStock.getMaxEffort(speed);
-        double rollingResistance = rollingStock.getRollingResistance(speed);
-        double weightForce = getWeightForce(rollingStock, path, position);
+        double grade = getAverageGrade(rollingStock, path, position);
+        double weightForce = getWeightForce(rollingStock, grade);
 
         if (action == Action.ACCELERATE)
-            tractionForce = maxTractionForce;
-
-        if (action == Action.BRAKE)
-            brakingForce = rollingStock.getMaxBrakingForce(speed);
+            tractionForce = rollingStock.getMaxEffort(speed);
+
+        if (action == Action.BRAKE) {
+            switch (useCase) {
+                case RUNNING_TIME -> brakingForce = rollingStock.getMaxBrakingForce(speed);
+                case ETCS_EBD -> {
+                    brakingForce = rollingStock.getSafeBrakingForce(speed);
+                    grade = getLowestGrade(rollingStock, path, position);
+                    weightForce = getWeightForce(rollingStock, grade);
```

Comment:
It feels weird to use the lowest grade only, but I assume it's how those curves are meant to be computed

Reply:
Yes exactly, that's how they're supposed to be computed, hence this new method